rusk/src/lib/lib.rs

```diff
@@ -133,80 +117,68 @@ impl Rusk {
 
         let mut dusk_spent = 0;
 
-        let mut nullifiers = BTreeSet::new();
-
-        // Here we discard transactions that:
-        // - Use nullifiers that are already use by previous TXs
-        // - Fail for any reason other than out of gas due to hitting the block
-        //   gas limit
-        'tx_loop: for tx in txs {
-            for nullifier in &tx.nullifiers {
-                if !nullifiers.insert(*nullifier) {
-                    discarded_txs.push(tx);
-                    continue 'tx_loop;
-                }
-            }
-
-            // The gas limit set for a transaction is either the limit it sets,
-            // or the gas left in the block, whichever is smallest.
-            let gas_limit = compute_gas_limit(tx.fee.gas_limit, block_gas_left);
-            session.set_point_limit(gas_limit);
-
-            let (gas_spent, call_result): (
-                u64,
-                Option<Result<RawResult, ContractError>>,
-            ) = match session.call(TRANSFER_CONTRACT, "execute", &tx) {
-                Ok(call_result) => call_result,
+        for tx in txs {
+            let (call_result, gas_spent) = match execute(
+                &mut session,
+                &tx,
+                block_gas_left,
+            ) {
+                // We're currently ignoring the result of a call.
+                Ok((_ret, gas_spent)) => (None, gas_spent),
                 Err(err) => match err {
-                    piecrust::Error::OutOfPoints => {
-                        let fee_limit =
-                            compute_gas_limit(tx.fee.gas_limit, u64::MAX);
-
-                        // If the transaction would have been out of points
-                        // with its own gas limit, it is invalid and should
-                        // be discarded.
-                        if gas_limit == fee_limit {
-                            discarded_txs.push(tx);
-                        }
+                    // An unspendable transaction should be discarded
+                    TxError::Unspendable(_) => {
+                        discarded_txs.push(tx);
```

Comment:
Panic while spending requires a revert of the session, right? (since any query in the session change the state)

If so, you should discriminate between unspendable due to OOG or unspendable due to panic while spending.

Reply:
For the sake of info

```rust
    /// # Panics
    /// Any failure in the checks performed in processing the transaction will
    /// result in a panic. The contract expects the environment to roll back any
    /// change in state.
    pub fn spend(&mut self, tx: Transaction) {
```

Reply:
Panic while spending does *not* require a revert of the session. A panic during a call to `spend` gets reverted by `piecrust` itself, and we don't need to touch it.